zkvm/src/types.rs

```diff
@@ -329,21 +337,79 @@ impl Data {
             },
         }
     }
+
+    /// Encodes blinded Data values.
+    pub fn encode(&self, buf: &mut Vec<u8>) {
+        match self {
+            Data::Opaque(x) => {
+                buf.extend_from_slice(x);
+                return;
+            }
+            Data::Witness(w) => w.encode(buf),
+        };
+    }
+
+    /// Encodes blinded Data values with length.
+    /// LE32(len) || <bytes>
+    fn encode_with_len(&self, buf: &mut Vec<u8>) {
```

Comment:
I'd avoid factoring this out to keep code clearer - this length-prefix stuff is very specific to the frozencontract encoding, so unless we need to do this in 3 more places, i'd avoid having a separate method for this. WDYT?

Reply:
sounds good to me! I went with this route since we want `to_bytes` to consume `self` but the encode functions all take `&self` — it would be an easy switch to go back to having `to_bytes` take a reference, though!